cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -183,9 +183,19 @@ static inline Future<std::shared_ptr<csv::StreamingReader>> OpenReaderAsync(
   auto tracer = arrow::internal::tracing::GetTracer();
   auto span = tracer->StartSpan("arrow::dataset::CsvFileFormat::OpenReaderAsync");
 #endif
+  ARROW_ASSIGN_OR_RAISE(
+      auto fragment_scan_options,
+      GetFragmentScanOptions<CsvFragmentScanOptions>(
+          kCsvTypeName, scan_options.get(), format.default_fragment_scan_options));
   ARROW_ASSIGN_OR_RAISE(auto reader_options, GetReadOptions(format, scan_options));
-
   ARROW_ASSIGN_OR_RAISE(auto input, source.OpenCompressed());
+  if (reader_options.encoding != "UTF-8") {
+    if (fragment_scan_options->stream_transform_func) {
+      ARROW_ASSIGN_OR_RAISE(input, fragment_scan_options->stream_transform_func(input));
+    } else {
+      return Status::Invalid("File encoding is not UTF-8, but no stream_transform_func has been provided.");
```

Comment:
Invalid is fine (IOError doesn't quite fit IMO)

```suggestion
      return Status::Invalid("File encoding is not UTF-8, but no stream_transform_func has been provided");
```